File: lig/src/test/kotlin/dev/ligature/lig/LigSuite.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.lig

import arrow.core.Either
import dev.ligature.*
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import dev.ligature.rakkoon.*

class LigSuite : FunSpec() {
    init {
        val ligParser = LigParser()
        val ligWriter = LigWriter()

        test("write entities") {
            val entity = Entity("test")
            val res = ligWriter.writeEntity(entity)
            res shouldBe "<test>"
        }

        test("parse entities") {
            val test = "<test>"
            val entity = ligParser.parseEntity(Rakkoon(test), null)
            entity shouldBe Either.Right(Entity("test"))
        }

        test("write attributes") {
            val attribute = Attribute("test")
            val res = ligWriter.writeAttribute(attribute)
            res shouldBe "@<test>"
        }

        test("parse attributes") {
            val test = "@<test>"
            val attribute = ligParser.parseAttribute(Rakkoon(test), null)
            attribute shouldBe Either.Right(Attribute("test"))
        }

        test("write FloatLiteral") {
            val test = FloatLiteral(3.0)
            val res = ligWriter.writeValue(test)
            res shouldBe "3.0"
        }

        test("parse FloatLiteral") {
            val test = "3.5"
            val res = ligParser.parseFloatLiteral(Rakkoon(test))
            res shouldBe Either.Right(FloatLiteral(3.5))
        }

        test("write IntegerLiteral") {
            val test = IntegerLiteral(3535)
            val res = ligWriter.writeValue(test)
            res shouldBe "3535"
        }

        test("parse IntegerLiteral") {
            val test = "3452345"
            val res = ligParser.parseIntegerLiteral(Rakkoon(test))
            res shouldBe Either.Right(IntegerLiteral(3452345))
        }

        test("write StringLiteral") {
            val test = StringLiteral("3535 55Hello")
            val res = ligWriter.writeValue(test)
            res shouldBe "\"3535 55Hello\""
        }

        test("parse StringLiteral") {
            val test = "\"3452345\\nHello\""
            val res = ligParser.parseStringLiteral(Rakkoon(test))
            res shouldBe Either.Right(StringLiteral("3452345\\nHello"))
        }

        test("basic Statement with all Entities") {
            val statement = Statement(Entity("e1"), Attribute("a1"), Entity("e2"), Entity("context"))
            val lines = ligWriter.write(listOf(statement).iterator())
            val resStatements = ligParser.parse(lines)
            listOf(statement) shouldBe resStatements.asSequence().toList()
        }

        test("list of Statements with Literal Values") {
            val statements = listOf(
                Statement(Entity("e1"), Attribute("a1"), Entity("e2"), Entity("context")),
                Statement(Entity("e2"), Attribute("a2"), StringLiteral("string literal"), Entity("context2")),
                Statement(Entity("e2"), Attribute("a3"), IntegerLiteral(Long.MAX_VALUE), Entity("context3")),
                Statement(Entity("e3"), Attribute("a4"), FloatLiteral(7.5), Entity("context4"))
            )
            val lines = ligWriter.write(statements.iterator())
            val resStatements = ligParser.parse(lines)
            statements shouldBe resStatements.asSequence().toList()
        }

//        test("parsing with wildcards") {
//            val textInput = "<e1> @<a2> 777 <e5>\n" +
//                    "_ @<a3> _ <e6>\n" +
//                    "_ _ \"Hello\" _\n" +
//                    "<e7> _ _ <e5>\n"
//            val expectedStatements = listOf(
//                Statement(Entity("e1"), Attribute("a2"), IntegerLiteral(777), Entity("e5")),
//                Statement(Entity("e1"), Attribute("a3"), IntegerLiteral(777), Entity("e6")),
//                Statement(Entity("e1"), Attribute("a3"), StringLiteral("Hello"), Entity("e6")),
//                Statement(Entity("e7"), Attribute("a3"), StringLiteral("Hello"), Entity("e5"))
//            )
//            val resStatements = ligParser.parse(textInput)
//            expectedStatements shouldBe resStatements.asSequence().toList()
//        }
    }
}

File: lig/src/test/scala/dev/ligature/lig/LigSuite.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.lig

import munit.FunSuite
import dev.ligature.{Identifier, LigatureLiteral, Statement}
import dev.ligature.gaze.Gaze

class LigSuite extends FunSuite {
  val testIdentifier = Identifier.fromString("test").getOrElse(???)
  def identifier(id: String) = Identifier.fromString(id).getOrElse(???)

  test("basic Statement with all Entities") {
    val statement = Statement(
      identifier("e1"),
      identifier("a1"),
      identifier("e2")
    )
    val lines = write(List(statement).iterator)
    val resStatements = read(lines)
    resStatements match {
      case Right(statements) => assertEquals(statements, List(statement))
      case Left(err)         => fail("failed", clues(err))
    }
  }

  test("list of Statements with Literal Values") {
    val statements = List(
      Statement(
        identifier("e1"),
        identifier("a1"),
        identifier("e2")
      ),
      Statement(
        identifier("e2"),
        identifier("a2"),
        LigatureLiteral.StringLiteral("string literal")
      ),
      Statement(
        identifier("e2"),
        identifier("a3"),
        LigatureLiteral.IntegerLiteral(Long.MaxValue)
      )
    )
    val lines = write(statements.iterator)
    val resStatements = read(lines)
    resStatements match {
      case Right(res) => assertEquals(res, statements)
      case Left(err)  => fail("failed", clues(err))
    }
  }

  test("parse Statement from multi-line String") {
    val statements =
      """
        |<a> <b> <c>
        |<a> <b> 123
        |<a> <b> "Test"
        |""".stripMargin
    val expectedStatements = Set(
      Statement(identifier("a"), identifier("b"), identifier("c")),
      Statement(identifier("a"), identifier("b"), LigatureLiteral.IntegerLiteral(123)),
      Statement(identifier("a"), identifier("b"), LigatureLiteral.StringLiteral("Test"))
    )
    val resStatements = read(statements)
    resStatements match {
      case Right(res) => assertEquals(res.toSet, expectedStatements)
      case Left(err)  => fail("failed", clues(err))
    }
  }

  test("parse identifiers") {
    val test = "<test>"
    val identifier = parseIdentifier(Gaze.from(test), Map(), None)
    assertEquals(identifier, Right(testIdentifier))
  }

  test("parse complex identifier") {
    val identifierS = "<http$://&&this@2]34.[42;342?#--__>"
    val identifierRes = parseIdentifier(Gaze.from(identifierS), Map(), None)
    assertEquals(
      identifierRes,
      Right(identifier("http$://&&this@2]34.[42;342?#--__"))
    )
  }

  test("parse IntegerLiteral") {
    val test = "3452345"
    val res: Either[LigError, LigatureLiteral] = parseIntegerLiteral(Gaze.from(test))
    assertEquals(res, Right(LigatureLiteral.IntegerLiteral(3452345)))
  }

  test("parse StringLiteral") {
    val test = "\"3452345\\nHello\""
    val res: Either[LigError, LigatureLiteral] = parseStringLiteral(Gaze.from(test))
    assertEquals(res, Right(LigatureLiteral.StringLiteral("3452345\\nHello")))
  }

  test("write identifiers") {
    val res = writeIdentifier(testIdentifier)
    assertEquals(res, "<test>")
  }

  test("write IntegerLiteral") {
    val test = LigatureLiteral.IntegerLiteral(3535)
    val res = writeValue(test)
    assertEquals(res, "3535")
  }

  test("write StringLiteral") {
    val test = LigatureLiteral.StringLiteral("3535 55Hello")
    val res = writeValue(test)
    assertEquals(res, "\"3535 55Hello\"")
  }
}
